### PR TITLE
fix(mcp): OAuth login fails with 400 when SSO returns id_token in URL fragment

### PR DESCRIPTION
**Pull request by Vishnuujain**
Issue: https://github.com/open-metadata/OpenMetadata/issues/29229

## Fixes

**`McpCallbackServlet`** — serve a JS fragment-extraction page instead of returning 400 when both `state` and `id_token` query params are absent. The page reads `window.location.hash`, extracts `id_token`, and submits it via a hidden form POST (not a GET redirect, per RFC 6819 §5.3.5) so the token never appears in a URL. Added `doPost()` to receive the form submission and delegate to `handleDirectIdTokenFlow()`.

**`UserSSOOAuthProvider`** — check `response.isCommitted()` before throwing `AuthorizeException` on missing pac4j state; return `SSO_REDIRECT_INITIATED` for the active-session path where no pac4j state is expected.

**`OAuthHttpStatelessServerTransportProvider`** — guard `sendRedirect()` with `response.isCommitted()` check to prevent `IllegalStateException: Committed` when the SSO provider already committed the response.

Also adds structured debug logging throughout the MCP auth flow (request params, response-committed state, session attribute dump on failure) to make future auth issues easier to diagnose.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes a 400 error during MCP OAuth login when an SSO provider (e.g., Google, Azure with an active session) returns the `id_token` in the URL fragment (`#id_token=...`) rather than as a query parameter — fragments are client-side only and the server never receives them.

- **`McpCallbackServlet`**: when no `state` or `id_token` query params are present, serves a minimal JS page that reads `window.location.hash`, extracts the `id_token`, and submits it via hidden form POST (avoiding URL/log exposure per RFC 6819 §5.3.5); `doPost()` validates the `Origin` header for CSRF protection and delegates to the existing `handleDirectIdTokenFlow()`.
- **`UserSSOOAuthProvider`**: skips the `AuthorizeException` when `response.isCommitted()`, recognising that `handleLogin()` already committed a direct redirect in the active-session shortcut path.
- **`OAuthHttpStatelessServerTransportProvider`**: guards `sendRedirect()` with `response.isCommitted()` to prevent `IllegalStateException` when the SSO provider has already committed the response.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge; the fix is well-scoped and the CSRF guard correctly rejects cross-origin requests, including the previously-identified case where the server origin cannot be resolved.

All three layers of the fix (fragment-extraction page, active-session shortcut detection, committed-response guard) cooperate correctly. CSRF protection via Origin validation now fails-closed when the server origin is unknown. The only weakness is that the port-stripping unit tests exercise an inline copy of the production logic rather than the actual method, which is a test-quality concern and does not affect runtime behaviour.

McpCallbackServletTest.java — the resolveServerOrigin port-stripping tests cover a reimplemented copy; the actual McpCallbackServlet.resolveServerOrigin() path is not exercised by those tests.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java | Core change: adds doPost() for form-based id_token submission, serves a JS fragment-extraction page instead of returning 400 when no state/id_token params are present, and adds CSRF Origin checking. Logic is sound; CSRF null-origin fallback now correctly rejects rather than allowing when the server origin is unknown. |
| openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/provider/UserSSOOAuthProvider.java | Guards the AuthorizeException throw with a response.isCommitted() check so the active-session shortcut (SSO committed a direct redirect) no longer surfaces as a server error; adds debug logging. Change is minimal and targeted. |
| openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/transport/OAuthHttpStatelessServerTransportProvider.java | Adds an isCommitted() guard before calling sendRedirect() so the IllegalStateException no longer fires when the SSO provider already committed the response. The sanitizeRedirectUrlForLogging helper strips query params before logging. |
| openmetadata-mcp/src/test/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServletTest.java | Good coverage of doPost edge cases, CSRF checks, and fragment-extraction page properties. The port-stripping tests exercise a reimplemented copy of the production logic rather than the actual resolveServerOrigin() method. |
| openmetadata-mcp/src/test/java/org/openmetadata/mcp/server/transport/OAuthHttpStatelessServerTransportProviderTest.java | New test file covering sanitizeRedirectUrlForLogging; correctly notes the committed-response guard needs integration tests and does not attempt to fake the full auth stack. |

</details>

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant B as Browser
    participant MCP as McpCallbackServlet
    participant SSO as SSO Provider
    participant UP as UserSSOOAuthProvider
    participant TR as OAuthHttpStatelessServerTransportProvider

    B->>TR: GET /mcp/authorize
    TR->>UP: handleAuthorizeRequest()
    UP->>SSO: handleLogin() (pac4j)
    SSO-->>B: "302 redirect /mcp/callback#id_token=eyJ... (active session)"
    Note over UP: response.isCommitted() → skip AuthorizeException
    Note over TR: response.isCommitted() → skip sendRedirect()

    B->>MCP: GET /mcp/callback (no query params, fragment is client-only)
    MCP-->>B: 200 HTML+JS fragment-extraction page
    Note over B: JS reads window.location.hash, extracts id_token

    B->>MCP: "POST /mcp/callback (form body: id_token=eyJ...)"
    Note over MCP: isOriginAllowed() checks Origin header (CSRF)
    MCP->>MCP: handleDirectIdTokenFlow()
    MCP-->>B: 302 → MCP redirect_uri (auth complete)
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant B as Browser
    participant MCP as McpCallbackServlet
    participant SSO as SSO Provider
    participant UP as UserSSOOAuthProvider
    participant TR as OAuthHttpStatelessServerTransportProvider

    B->>TR: GET /mcp/authorize
    TR->>UP: handleAuthorizeRequest()
    UP->>SSO: handleLogin() (pac4j)
    SSO-->>B: "302 redirect /mcp/callback#id_token=eyJ... (active session)"
    Note over UP: response.isCommitted() → skip AuthorizeException
    Note over TR: response.isCommitted() → skip sendRedirect()

    B->>MCP: GET /mcp/callback (no query params, fragment is client-only)
    MCP-->>B: 200 HTML+JS fragment-extraction page
    Note over B: JS reads window.location.hash, extracts id_token

    B->>MCP: "POST /mcp/callback (form body: id_token=eyJ...)"
    Note over MCP: isOriginAllowed() checks Origin header (CSRF)
    MCP->>MCP: handleDirectIdTokenFlow()
    MCP-->>B: 302 → MCP redirect_uri (auth complete)
```

</a>
</details>

<sub>Reviews (7): Last reviewed commit: ["refactor(mcp): string constants + lazy-c..."](https://github.com/open-metadata/openmetadata/commit/1c4dd3315763b90ebc971d22bee5d057c70fb0a7) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=38684004)</sub>

### CODE REVIEW

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java [hunk 1]
```diff
@@ -217,6 +217,35 @@ private IdTokenValidator createIdTokenValidator(AuthenticationCodeFlowHandler ss
...
+      throws ServletException, IOException {
+    // Handles the form POST from serveFragmentExtractionPage() — the JS page extracts the
+    // id_token from window.location.hash and submits it here via a hidden form so the token
+    // never appears in a URL, browser history, access logs, or Referer headers.
+    AuthenticationCodeFlowHandler ssoHandler = resolveSsoHandler();
+    if (ssoHandler == null) {
+      response.sendError(
+          HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+          "MCP SSO not available. Please restart the server.");
+      return;
```
Comment (harshach) [inline openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java]: can we do if else, instead of using “return” as the handler here

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/transport/OAuthHttpStatelessServerTransportProvider.java [hunk 2]
```diff
@@ -974,4 +982,12 @@ private void sendOAuthError(
     body.put("error_description", description);
     getObjectMapper().writeValue(response.getOutputStream(), body);
   }
+
+  private static String sanitizeRedirectUrlForLogging(String url) {
```
Comment (harshach) [inline openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/transport/OAuthHttpStatelessServerTransportProvider.java:986]: we needd to have unit tests for this

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java [hunk 3]
```diff
@@ -217,6 +224,100 @@ private IdTokenValidator createIdTokenValidator(AuthenticationCodeFlowHandler ss
...
+          HttpServletResponse.SC_SERVICE_UNAVAILABLE,
+          "MCP SSO not available. Please restart the server.");
+    } else if (!isOriginAllowed(request)) {
+      // CSRF protection: the form is always submitted from the same origin as this server.
+      // Reject any POST whose Origin header does not match the server's own base URL so a
+      // malicious cross-origin page cannot submit an attacker-controlled id_token to hijack
+      // a victim's pending MCP auth session.
+      LOG.warn(
+          "MCP OAuth doPost rejected: Origin '{}' does not match server origin",
+          request.getHeader("Origin"));
```
Comment (harshach) [inline openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java:257]: All of these string literals should be final constants

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java [hunk 4]
```diff
@@ -217,6 +224,100 @@ private IdTokenValidator createIdTokenValidator(AuthenticationCodeFlowHandler ss
...
+      response.sendError(
+          HttpServletResponse.SC_FORBIDDEN,
+          "CSRF protection: request origin does not match server origin");
+    } else {
+      String idTokenParam = request.getParameter("id_token");
```
Comment (harshach) [inline openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java:260]: Same

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java [hunk 5]
```diff
@@ -217,6 +224,100 @@ private IdTokenValidator createIdTokenValidator(AuthenticationCodeFlowHandler ss
...
+    }
+    return origin.equals(serverOrigin);
+  }
+
+  String resolveServerOrigin() {
+    try {
+      MCPConfiguration mcpConfig = SecurityConfigurationManager.getCurrentMcpConfig();
+      String baseUrl = mcpConfig != null ? mcpConfig.getBaseUrl() : null;
+      if (baseUrl == null) {
+        SystemRepository systemRepo = Entity.getSystemRepository();
```
Comment (harshach) [inline openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java:314]: any reaosn we are not doing this as part of Class init, that way we cache it or reduce the db lookups. which means the url config when it chagned via UI setttings page requires server restart. We should think about the db calls , how many times are we going to resolve the server origin,

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java (hunk 3 above)
Comment (Vishnuujain) [reply openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java:257]: fixed

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java (hunk 5 above)
Comment (Vishnuujain) [reply openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java:314]: makes sense, added caching to reduce db lookup

Comment (pmbrull) [reply openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java:314]: this should still be moved to the class init no? or how many times are we calling the `resolveServerOrigin` method?

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java (hunk 1 above)
Comment (Vishnuujain) [reply openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java]: done

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/transport/OAuthHttpStatelessServerTransportProvider.java (hunk 2 above)
Comment (Vishnuujain) [reply openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/transport/OAuthHttpStatelessServerTransportProvider.java:986]: done

File: openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java (hunk 5 above)
Comment (Vishnuujain) [reply openmetadata-mcp/src/main/java/org/openmetadata/mcp/server/auth/handlers/McpCallbackServlet.java:314]: moving to constructor init risks permanently caching null if the DB isn't populated at server startup and now resolveServerOrigin is called at most once per server lifetime

**Review by pmbrull — APPROVED** [review]
